subprojects/execution/src/main/java/org/gradle/internal/execution/steps/BuildOperationStep.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps;
+
+import org.gradle.internal.execution.Context;
+import org.gradle.internal.execution.Result;
+import org.gradle.internal.execution.Step;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationDescriptor;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.CallableBuildOperation;
+
+import java.util.function.Function;
+
+public abstract class BuildOperationStep<C extends Context, R extends Result> implements Step<C, R> {
+    private final BuildOperationExecutor buildOperationExecutor;
+
+    protected BuildOperationStep(BuildOperationExecutor buildOperationExecutor) {
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    protected <T> T operation(Function<BuildOperationContext, T> operation, BuildOperationDescriptor.Builder description) {
```

Comment:
I think it would be nicer to have the operation as the last argument, since it is the actual action which is executed.

Reply:
So that would change:

```java
        return operation(operationContext -> {
                BeforeExecutionState beforeExecutionState = captureExecutionState(executionContext);
                operationContext.setResult(Operation.Result.INSTANCE);
                return beforeExecutionState;
            },
            BuildOperationDescriptor
                .displayName("Snapshot inputs and outputs before executing " + executionContext.getWork().getDisplayName())
                .details(Operation.Details.INSTANCE)
        );
```

into:

```java
        return operation(BuildOperationDescriptor
                .displayName("Snapshot inputs and outputs before executing " + executionContext.getWork().getDisplayName())
                .details(Operation.Details.INSTANCE),
            operationContext -> {
                BeforeExecutionState beforeExecutionState = captureExecutionState(executionContext);
                operationContext.setResult(Operation.Result.INSTANCE);
                return beforeExecutionState;
            }
        );
```

I'm not entirely sure about which I prefer. I like that the `return` statement is closer to what is actually returning the value in the first case. If creating the descriptor was a simpler affair I'd agree with you, but since it requires multiple lines of code, I think I'd like to keep the current setup.